src/source/PeerConnection/SessionDescription.c

```diff
@@ -316,6 +326,65 @@ PCHAR fmtpForPayloadType(UINT64 payloadType, PSessionDescription pSessionDescrip
     return NULL;
 }
 
+/*
+ * Extracts a (hex) value after the provided prefix string. Returns true if
+ * successful.
+ */
+static BOOL readHexValue(PCHAR input, PCHAR prefix, PINT32 value) {
+  PCHAR substr = STRSTR(input, prefix);
+  if (substr != NULL) {
+    if (sscanf(substr + STRLEN(prefix), "%x", value) == 1) {
+      return TRUE;
+    }
+  }
+  return FALSE;
+}
+
+/*
+ * Scores the provided fmtp string based on this library's ability to
+ * process various types of H264 streams. A score of 0 indicates an
+ * incompatible fmtp line. Beyond this, a higher score indicates more
+ * compatibility with the desired characteristics, packetization-mode=1,
+ * level-asymmetry-allowed=1, and inbound match with our preferred
+ * profile-level-id. Scoring is on a scale from [0,1], expressed as the
+ * inverse of fitness distance (1 / fitness-distance) as defined here:
+ * https://www.w3.org/TR/mediacapture-streams/#dfn-fitness-distance
+ */
+DOUBLE getH264FmtpScore(PCHAR fmtp) {
+  INT32 profileId = 0, packetizationMode = 0, levelAsymmetry = 0;
+  BOOL isAsymmetryAllowed = FALSE;
+  BOOL isProfileMatch = FALSE;
+
+  // No ftmp match found.
+  if (fmtp == NULL) {
+      return 0.;
+  }
+
+  // Currently, the packetization mode must be 1, as the packetization logic
+  // is currently not configurable, and sends both NALU and FU-A packets.
+  // https://tools.ietf.org/html/rfc7742#section-6.2
+  if (!readHexValue(fmtp, "packetization-mode=", &packetizationMode) || packetizationMode != 1) {
+      return 0.;
+  }
+
+  if (readHexValue(fmtp, "profile-level-id=", &profileId)) {
+      isProfileMatch = (profileId == H264_PROFILE_42E01F);
```

Comment:
Some browsers do not have the `profile-level-id` of `42E01F`. The `profile-level-id` consists of three parts, which are `profile_idc (8 bits)`, `profile_iop (8 bits)`, and `level_idc (8 bits)`.

So I think it should be more DETAILED when calculating the fmtp score. The score of `profile_idc` of `0x42` must be higher than `0x64`.

Reply:
This is a good suggestion, however, for the context of this PR I would leave it out and follow-up with another change to determine the profile-level-id. This is a terrible RFC all together and mangling different meanings in a free-handed text is a bad solution. At any rate, we should take each one of these individually as supporting the entire RFC makes no sense - just the parts that are real-life use cases which these actually are.

Reply:
@docee Thanks! It's a good point. Have you come across any SDP offers that don't support `42E01F`? If so, could you send me a sample? I'm trying to build a small library of real-world SDP offers in this PR so that the SDK doesn't unknowingly exclude a previously supported platform with a change. 

I'd like to try and get this in ASAP (I've already let it linger too long) as the current code config'd to use H264 seems to be broken "as-is" on several major browsers/platforms. But I see what you're saying and we probably ought to follow up by somehow generalizing H264 support beyond `RTC_CODEC_H264_PROFILE_42E01F_LEVEL_ASYMMETRY_ALLOWED_PACKETIZATION_MODE`. 

As @MushMal pointed out, the RFC is both broad and convoluted, so it seems like achieving broad compatibility is much more dependent on finding and supporting the most commonly implemented encoding schemes than it is about satisfying the full breadth of what's offered in the spec. I'm definitely nervous that this change "misses" a sweet spot among other devices.

The support for transceiver configuration in this SDK is (as you noted) very constrained, as is support for any sort of asymmetrical encoding configuration. For example, it doesn't even expose what's in the offer to the API user, so there's no opportunity (for example) to reconfigure encoding based on available options. That's certainly an area for growth!

Reply:
This is Chrome SDP on android:
```
"v=0
o=- 2383219383355379692 2 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE 0 1 2
a=msid-semantic: WMS 3CXV4snScv28Bl5Ltn7V4StSDzTGKOnaaAdf
m=audio 9 UDP/TLS/RTP/SAVPF 111 103 9 0 8 105 13 110 113 126
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:Wafl
a=ice-pwd:IS933a+6pookL48tYXASXnwW
a=ice-options:trickle
a=fingerprint:sha-256 49:3C:77:6E:6B:2B:90:69:00:AE:47:4A:87:A7:F4:F6:F0:B3:6D:D6:FA:7F:84:DD:A5:6E:6E:2E:1F:64:D7:7F
a=setup:actpass
a=mid:0
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=extmap:2 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=extmap:4 urn:ietf:params:rtp-hdrext:sdes:mid
a=extmap:5 urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id
a=extmap:6 urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id
a=sendrecv
a=msid:3CXV4snScv28Bl5Ltn7V4StSDzTGKOnaaAdf f94c5ff6-26b9-4315-815d-40b4dd2efdef
a=rtcp-mux
a=rtpmap:111 opus/48000/2
a=rtcp-fb:111 transport-cc
a=fmtp:111 minptime=10;useinbandfec=1
a=rtpmap:103 ISAC/16000
a=rtpmap:9 G722/8000
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:105 CN/16000
a=rtpmap:13 CN/8000
a=rtpmap:110 telephone-event/48000
a=rtpmap:113 telephone-event/16000
a=rtpmap:126 telephone-event/8000
a=ssrc:2261433415 cname:5L7rDchCDtUnZEok
a=ssrc:2261433415 msid:3CXV4snScv28Bl5Ltn7V4StSDzTGKOnaaAdf f94c5ff6-26b9-4315-815d-40b4dd2efdef
a=ssrc:2261433415 mslabel:3CXV4snScv28Bl5Ltn7V4StSDzTGKOnaaAdf
a=ssrc:2261433415 label:f94c5ff6-26b9-4315-815d-40b4dd2efdef
m=video 9 UDP/TLS/RTP/SAVPF 96 97 98 99 127 125 104 122 106 107 108 109 124 121 100 101 102
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:Wafl
a=ice-pwd:IS933a+6pookL48tYXASXnwW
a=ice-options:trickle
a=fingerprint:sha-256 49:3C:77:6E:6B:2B:90:69:00:AE:47:4A:87:A7:F4:F6:F0:B3:6D:D6:FA:7F:84:DD:A5:6E:6E:2E:1F:64:D7:7F
a=setup:actpass
a=mid:1
a=extmap:14 urn:ietf:params:rtp-hdrext:toffset
a=extmap:2 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:13 urn:3gpp:video-orientation
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=extmap:12 http://www.webrtc.org/experiments/rtp-hdrext/playout-delay
a=extmap:11 http://www.webrtc.org/experiments/rtp-hdrext/video-content-type
a=extmap:7 http://www.webrtc.org/experiments/rtp-hdrext/video-timing
a=extmap:8 http://www.webrtc.org/experiments/rtp-hdrext/color-space
a=extmap:4 urn:ietf:params:rtp-hdrext:sdes:mid
a=extmap:5 urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id
a=extmap:6 urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id
a=recvonly
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:96 VP8/90000
a=rtcp-fb:96 goog-remb
a=rtcp-fb:96 transport-cc
a=rtcp-fb:96 ccm fir
a=rtcp-fb:96 nack
a=rtcp-fb:96 nack pli
a=rtpmap:97 rtx/90000
a=fmtp:97 apt=96
a=rtpmap:98 VP9/90000
a=rtcp-fb:98 goog-remb
a=rtcp-fb:98 transport-cc
a=rtcp-fb:98 ccm fir
a=rtcp-fb:98 nack
a=rtcp-fb:98 nack pli
a=fmtp:98 profile-id=0
a=rtpmap:99 rtx/90000
a=fmtp:99 apt=98
a=rtpmap:127 VP9/90000
a=rtcp-fb:127 goog-remb
a=rtcp-fb:127 transport-cc
a=rtcp-fb:127 ccm fir
a=rtcp-fb:127 nack
a=rtcp-fb:127 nack pli
a=fmtp:127 profile-id=1
a=rtpmap:125 VP9/90000
a=rtcp-fb:125 goog-remb
a=rtcp-fb:125 transport-cc
a=rtcp-fb:125 ccm fir
a=rtcp-fb:125 nack
a=rtcp-fb:125 nack pli
a=fmtp:125 profile-id=2
a=rtpmap:104 H264/90000
a=rtcp-fb:104 goog-remb
a=rtcp-fb:104 transport-cc
a=rtcp-fb:104 ccm fir
a=rtcp-fb:104 nack
a=rtcp-fb:104 nack pli
a=fmtp:104 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=420015
a=rtpmap:122 rtx/90000
a=fmtp:122 apt=104
a=rtpmap:106 H264/90000
a=rtcp-fb:106 goog-remb
a=rtcp-fb:106 transport-cc
a=rtcp-fb:106 ccm fir
a=rtcp-fb:106 nack
a=rtcp-fb:106 nack pli
a=fmtp:106 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=4d0015
a=rtpmap:107 rtx/90000
a=fmtp:107 apt=106
a=rtpmap:108 H264/90000
a=rtcp-fb:108 goog-remb
a=rtcp-fb:108 transport-cc
a=rtcp-fb:108 ccm fir
a=rtcp-fb:108 nack
a=rtcp-fb:108 nack pli
a=fmtp:108 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=640015
a=rtpmap:109 rtx/90000
a=fmtp:109 apt=108
a=rtpmap:124 H264/90000
a=rtcp-fb:124 goog-remb
a=rtcp-fb:124 transport-cc
a=rtcp-fb:124 ccm fir
a=rtcp-fb:124 nack
a=rtcp-fb:124 nack pli
a=fmtp:124 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e015
a=rtpmap:121 rtx/90000
a=fmtp:121 apt=124
a=rtpmap:100 red/90000
a=rtpmap:101 rtx/90000
a=fmtp:101 apt=100
a=rtpmap:102 ulpfec/90000
m=application 9 UDP/DTLS/SCTP webrtc-datachannel
c=IN IP4 0.0.0.0
a=ice-ufrag:Wafl
a=ice-pwd:IS933a+6pookL48tYXASXnwW
a=ice-options:trickle
a=fingerprint:sha-256 49:3C:77:6E:6B:2B:90:69:00:AE:47:4A:87:A7:F4:F6:F0:B3:6D:D6:FA:7F:84:DD:A5:6E:6E:2E:1F:64:D7:7F
a=setup:actpass
a=mid:2
a=sctp-port:5000
a=max-message-size:262144
"
```

Reply:
Thanks @docee! I think @lherman-cs and @MushMal moved over to another PR (they were going to push some final changes) - I'll make sure that I follow up and get your SDP capture in the test set!!!

Reply:
Thanks @docee! The SDP has been included in the unit test, https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/966/files#diff-560b56b36b72eba42e1ed20d819402995117a0cdab25953080198ca7c13e8613R1011-R1014